reduceSPI/nn.py

```diff
@@ -0,0 +1,890 @@
+        out_channels : int, number of channels.
+
+        Returns
+        -------
+        out_shape : tuple, shape of the output of the layer
+
+        """
+        return conv_output_size(
+            in_shape, out_channels,
+            kernel_size=self.enc_ks,
+            stride=self.enc_str,
+            padding=self.enc_pad,
+            dilation=self.enc_dil)
+
+    def forward(self, h):
+        """
+        Compute the passage through the neural network
+
+        Parameters
+        ----------
+        h : tensor, image or voxel.
+
+        Returns
+        -------
+        mu : tensor, latent space mu.
+        logvar : tensor, latent space sigma.
+
+        """
+        for i in range(self.n_blocks):
+            h = self.blocks[2*i](h)
+            h = self.blocks[2*i+1](h)
+            h = self.leakyrelu(h)
+        h = h.view(-1, self.fcs_infeatures)
+        mu = self.fc_mu(h)
+        logvar = self.fc_logvar(h)
+        if self.latent_space_definition == 0:
+            z = reparametrize(mu, logvar)
+            return mu, logvar, z, torch.zeros(mu.shape)
+        elif self.latent_space_definition in [1, 2, 3]:
+            rot_mats_no_var, translation, logvar_rotmat = reparametrize_so3(
+                mu, logvar)
+        else:
+            rot_mats_no_var, translation, logvar_rotmat = reparametrize_s2s2(
+                mu, logvar)
+        rot_mats = add_logvar(rot_mats_no_var, logvar_rotmat)
+        if self.latent_space_definition == 1:
+            vec_rot = rot_mats.reshape(-1, 9)
+            z = torch.cat((vec_rot, translation), dim=1)
+            return mu, logvar, z, rot_mats
+        wig_matrix = convert_rot_mat_ten_to_wig_mat_ten(
+            rot_mats, self.latent_dim)
+        z0 = torch.matmul(wig_matrix, self.z0)
+        z1 = torch.matmul(wig_matrix, self.z1)
+        z = torch.cat((z0, z1, translation), dim=1)
+        return mu, logvar, z, rot_mats
+
+
+def reparametrize_so3(mu, logvar):
+    mu_rotmat = mu.T[:-2].T
+    mu_transl = mu.T[-2:].T
+    logvar_rotmat = logvar.T[:-2].T
+    logvar_transl = logvar.T[-2:].T
+    rot_mats_no_var = rot_mat_tensor(mu_rotmat, 3)
+    translation = reparametrize(mu_transl, logvar_transl)
+    return rot_mats_no_var, translation, logvar_rotmat
+
+
+def reparametrize_s2s2(mu, logvar):
+    mu_rotmat = mu.T[:-2].T
+    mu_transl = mu.T[-2:].T
+    logvar_rotmat = logvar.T[:-2].T
+    logvar_transl = logvar.T[-2:].T
+    translation = reparametrize(mu_transl, logvar_transl)
+    rot_mats_no_var = rotation_vectors(mu_rotmat)
+    return rot_mats_no_var, translation, logvar_rotmat
+
+
+def rotation_vectors(z):
+    z = z.reshape((-1, 6))
+    u = z.T[:3].T
+    v = z.T[3:].T
+    e1 = nn.functional.normalize(u)
+    v2 = v - (e1*v).sum(-1, keepdim=True) * e1
+    e2 = nn.functional.normalize(v2)
+    e3 = torch.cross(e1, e2)
+    matrices = torch.cat([e1, e2, e3]). reshape((-1, 3, 3))
+    return matrices
+
+
+def map_to_lie_algebra(v):
```

Comment:
This function needs a unit-test, to make sure that we do not make a mistake here (I'm not a good unit-tester just by reading)

Reply:
def from_vector_to_skewsym_mat